lxd/storage/backend_lxd.go

```diff
@@ -171,8 +173,75 @@ func (b *lxdBackend) CreateInstanceFromCopy(inst Instance, src Instance, snapsho
 	return ErrNotImplemented
 }
 
+// createInstanceSymlink creates a symlink in the instance directory to the instance's mount path.
+func (b *lxdBackend) createInstanceSymlink(inst Instance, mountPath string) error {
+	symlinkPath := inst.Path()
+	if !shared.PathExists(symlinkPath) {
+		err := os.Symlink(mountPath, symlinkPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// CreateInstanceFromImage creates a new volume for an instance populated with the image requested.
 func (b *lxdBackend) CreateInstanceFromImage(inst Instance, fingerprint string, op *operations.Operation) error {
-	return ErrNotImplemented
+	logger := logging.AddContext(b.logger, log.Ctx{"project": inst.Project(), "instance": inst.Name()})
+	logger.Debug("CreateInstanceFromImage started")
+	defer logger.Debug("CreateInstanceFromImage finished")
+
+	volType, err := InstanceTypeToVolumeType(inst.Type())
+	if err != nil {
+		return err
+	}
+
+	revert := true
+	defer func() {
+		if !revert {
+			return
+		}
+		b.DeleteInstance(inst, op)
+	}()
+
+	filler := func(instanceMountPath string) error {
+		var tracker *ioprogress.ProgressTracker
+		if op != nil { // Not passed when being done as part of pre-migration setup.
+			metadata := make(map[string]interface{})
+			tracker = &ioprogress.ProgressTracker{
+				Handler: func(percent, speed int64) {
+					shared.SetProgressMetadata(metadata, "create_instance_from_image_unpack", "Unpack", percent, 0, speed)
+					op.UpdateMetadata(metadata)
+				}}
+		}
+
+		imagePath := shared.VarPath("images", fingerprint)
+
+		// tomp TODO currently passing false to isBlockBackend argument, which only affects
+		// user hint on how to increase disk space if not enough to unpack image. Ask about
+		// removing this argument entirely.
+		return ImageUnpack(imagePath, instanceMountPath, false, b.state.OS.RunningInUserNS, tracker)
```

Comment:
@stgraber I've got a question with this bit, the `ImageUnpack` take a boolean argument indicating whether or not the backend is block or not (but not a block device that we will be using for VM), the only use of this boolean is to alter the error message returned to the user when there isn't enough room in the volume to unpack the image. It adds to the error "(consider increasing your pool's volume.size)" to the error message if true is passed.

I was wondering whether we should just remove this or make it always return this part because it could also apply if using project quotas.

Reply:
Project quotas apply to the container, not to the image though.

That's why we only show it for block based storage as that's the only case where we will create something of a fixed size and possibly attempt to unpack something larger in it.

For all other cases, the image itself isn't size constrained, the container created from it is.

Reply:
@stgraber ah I see OK makes sense.